[MRG+2] Simple Restyle of Docs

vsoch opened the pull request:
This is a simple PR to change some of the colors / style of the pydicom docs. Specifically, including the following:

## Main Color Scheme
the main color scheme is modified to be blue / gold / white, which nicely still matches the read the docs theme (blue) and python (blue and gold). The color that was prevalent/dominant before (black) is greatly reduced so it's much brighter. I originally had the upper left square of the sidebar as blue, but ultimately removed it in favor of white.

![screenshot from 2017-07-17 22-39-20](https://user-images.githubusercontent.com/814322/28298529-bc91b3a4-6b41-11e7-8575-15ef19978edc.png)

As you can see, the logo for pydicom is added, and a favicon as well. Likely the favicon isn't necessary as the one from pydicom.github.io would be found, but in case they are built in isolation, it is good to include here. Speaking of the main site, they match quite well! I think I like this one better, actually, because it has much better formatting of code / content, along with better use of the entire page space.

![image](https://user-images.githubusercontent.com/814322/28298690-b8d5a670-6b42-11e7-9316-f7d2056a13f1.png)


## Sphinx Gallery
The buttons for the sphinx gallery are modified to be gold, with a blue hover effect. This changes the previous shadow/gradient thing... the dark yellow mustard yellow gradient was making web designers everywhere die a little inside every time they saw it, lol.

![screenshot from 2017-07-17 22-40-03](https://user-images.githubusercontent.com/814322/28298543-d7507a4a-6b41-11e7-803a-12761cde3f05.png)

and the hover effect is just - blue.

![screenshot from 2017-07-17 22-40-08](https://user-images.githubusercontent.com/814322/28298554-e8468ae2-6b41-11e7-941e-aaffabc608d8.png)

## Local Building Docker Image
I added a simple [Dockerfile](https://github.com/vsoch/pydicom/blob/docs/style/doc/Dockerfile) and [README.md](https://github.com/vsoch/pydicom/blob/docs/style/doc/README.md) to give instructions for creating an image to build the docs locally. @glemaitre - I know you were against this when I added it to this [previous PR](https://github.com/glemaitre/pydicom/pull/1), but I don't see any other logical way to build / test and work on docs locally without the (pretty complex) set of dependencies that I don't want to install on my system. I am going to be using this very regularly, and I think we should provide it for other developers like myself. It's one Dockerfile and a README, and I hope you think this is reasonable to add.

![screenshot from 2017-07-17 22-39-35](https://user-images.githubusercontent.com/814322/28298643-73f0cf1c-6b42-11e7-9121-45bafec8b28b.png)

vsoch commented:
pep8 robot, if you show your face, I am going to throw :tomato: at you!

vsoch commented:
huh, I didn't change anything related to dataset - any ideas why this is failing for travis?

glemaitre commented:
I am fine with style. It looks pretty good. For the docker file, I am still against. Adding pip install package1 in the makefile in the doc statement seems much simpler to me. And if it is going to fail for an install reason the error are still pretty understandable. I also assume that in general a user setup is environment once and he never has to build the doc. He will use the online doc or the doctring via is editor IDE. So IMHO this is still an over-engineered solution. 

vsoch commented:
The Docker image means that I (optionally) don't need to install the multiple dependencies:

```
pip install Sphinx==1.4
pip install sphinx_rtd_theme
pip install alabaster 
pip install sphinx_bootstrap_theme
pip install numpydoc
```

that I hit when I attempt `make html`. No, I don't want to install them on my local machine so adding to the make file to do so is not a solution that works for me. My workflow stops here:

```
# These two lines make the build a bit more lengthy, and the
# the embedding of images more robust
rm -rf _build/html/_images
#rm -rf _build/doctrees/
sphinx-build -b html -d _build/doctrees   . _build/html
Running Sphinx v1.3.1

Extension error:
Could not import extension sphinx_gallery.gen_gallery (exception: No module named sphinx_gallery.gen_gallery)
make: *** [html] Error 1
```

and even sooner than that if I don't have sphinx. Docker is a common, modern solution for many developers that has an entire purpose of me **not** needing to install every specific set of dependencies  and maintain multiple virtual environments for the software I happen to be working on. Further, I don't see the harm in giving people another option. Minimally they can ignore it.

massich commented:
+1 to the Theme. I love it. 
-1 to container (I still find them unnecessarily complicated)
travis: you have trouble depending of the dependencies. When only numpy or no numpy and pillow instead you are fine. Any other combination fails. This is something that is beeing worked in #407

vsoch commented:
so both of your solutions is to just install everything on your local machines? And you are not open to even the idea of giving someone another option?

mrbean-bremen commented:
I like the styling a lot - though I'm not the design kind of person. About docker - I don't really care. Working under Windows (don't ask), I use virtual environments anyway.

massich commented:
> so both of your solutions is to just install everything on your local machines? And you are not open to even the idea of giving someone another option?

That's not it, I'm not against containers. If you think that containers are necessary go for it. What i'm not sure is that they should be within the codebase. In anycase, what I would do is a PR for the Theme (which is awesome, BTW) and then another PR to add the container. They are two different things.

vsoch commented:
It's a chicken and egg deal, I wouldn't have been able to work on this at all without the container.

vsoch commented:
would you be less averse to the idea of adding the Dockerfile and README under build_tools? I would still want a separate README.md in the docs folder for some instruction / link to different options for building and developing the docs. Right now you have to know something about how sphinx works **and** be willing to install dependencies to even develop. I think if the Dockerfile / pointer to it helps just one person, the hefty addition of a 21 line text file (that can be safely ignored) is worth it ;)

glemaitre commented:
> would you be less averse to the idea of adding the Dockerfile and README under build_tools?

I would be inclined to create a repository in pydicom organization containing all docker file there.
It would allow the user to install the full set of tools or part of it and build doc and stuff like that.

> Right now you have to know something about how sphinx works and be willing to install dependencies to even develop.

I would assume that Python developers can be more familiar with Python tools than container at first.

glemaitre commented:
If really needed I could see a docker folder in the root directory but I more in favor of the previous solution with centralized docker files.

bastula commented:
If you look at most Python web apps (or some libraries) they will have two `requirements.txt` files. One for production and one for development.

Following https://github.com/audreyr/cookiecutter-pypackage (which I used to build my own library), it puts the user requirements in `setup.py` and the dev requirements in `requirements_dev.py`.

This is another option that may work without over complicating things. Also `requirements_dev.txt` could be referenced in a dockerfile, keeping the project DRY.

Sent from my iPhone

> On Jul 18, 2017, at 06:16, Guillaume Lemaitre <notifications@github.com> wrote:
> 
> If really needed I could see a docker folder in the root directory but I more in favor of the previous solution with centralized docker files.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

glemaitre commented:
> If you look at most Python web apps (or some libraries) they will have two `requirements.txt` files. One for production and one for development.

It was one of my proposal to use a requirement.txt

vsoch commented:
I'm used to a design where the Dockerfile lives with the dependencies that it requires / works for, but I'm also ok with compromising. The requirements.txt is a great idea! However it still doesn't solve the issue of developing without installing dependencies. As a compromise (making this option available but not putting the Dockerfile in the repo, which doesn't seem to be well liked) I moved the docker container to belong in the [docker-containers repo](https://hub.docker.com/r/pydicom/pydicom-docs/), and I updated the docs README to point to it. The Travis fail seems to be the same as before, so I think this PR should be close!

I also saw the forced update to master (I merged current master into my local branch just now), any idea what that was?

glemaitre commented:
The problem with Travis is due to the dependencies and test. Nothing related to the PR. 

massich commented:
`.travis.yml` and `travis/install.sh` should have no changes. A rebase to master is needed. I tried to do it myself and push it to the PR but I think I messed up somewhere. (I'll keep investigating)

vsoch commented:
rebased.

massich commented:
yep, thanks. I could pull out the PR, but the pushing end up creating a new branch. (which I'll delete now, thanks)

vsoch commented:
Just curious - what is the deployment strategy for docs? We have a few options (and likely one of these is setup):

  - have CI build, test, and deploy to github pages
  - have the docs served in the main branch, and we choose to render from master/docs folder
  - have the docs built and deployed elsewhere (eg read the docs)

it looks like the old version is still live: https://pydicom.github.io/pydicom/stable/

Is the plan to wait for the next release (and updated docs) and then change?

glemaitre commented:
‎The doc is build and deploy in gh-pages and available in pydicom.github.io/pydicom the advantage is that you have access to the stable and dev version. 

vsoch commented:
ok cool. So we merge into master, and then separately update gh-pages (manually) when it's ready?

glemaitre commented:
No this is automatically done by circle-ci using this [script](https://github.com/pydicom/pydicom/blob/master/build_tools/circle/push_doc.sh)

vsoch commented:
ok, so should this have happened for this merge?

mrbean-bremen commented:
check at https://pydicom.github.io/pydicom/dev/

vsoch commented:
ahh that's what I was looking for! And now I understand, dev == master and production is a release, distinction is here:

```

if [ "$CIRCLE_BRANCH" = "master" ]
then
    dir=dev
else
    # Strip off .X
    dir="${CIRCLE_BRANCH::-2}"
fi
```
looking good!! :D

mrbean-bremen commented:
Aren't you sitting at the west coast? Interesting working hours...

vsoch commented:
lol yeah yeah, I'm going back to sleep :) How can I sleep when there are things going on the internet?

![image](https://user-images.githubusercontent.com/814322/28360560-a9a1f53a-6c2a-11e7-98a0-3bf1b87fd274.png)

mrbean-bremen commented:
:D

glemaitre commented:
:D as a side note, dev would become stable (we have to set that thought with a symlink) at the time of the pypi release.
